tools/aot/aot_compile.bzl

```diff
@@ -3,29 +3,243 @@
+            and validates the outputs against PyTorch.
+        foo_torch_exporter:
+            py_binary that runs the torch_loader function to get the `TorchLoaderOutput`
+            (containing the PyTorch program and inputs), then calls the upstream
+            `fx.export_and_import` API to generate Torch dialect, and finally runs the
+            PyTorch program on reference inputs and saves the reference outputs (as .npz)
+            which will eventually be used for validation of the AOT compiled code.
+        foo_execute_test_generator:
+            py_binary that reads the reference tensors to infer the function signature
+            (rank, element type for each input/output tensor) and then materializes the
+            templatized parameters in `execute_test.template.cpp`.
+        gen_foo_mlir_torch:
+            genrule that invokes `foo_torch_exporter` and saves the torch dialect program
+            (*_torch.mlir).
+        gen_foo_mlir_tcp:
+            genrule that invokes `tcp-opt` to convert the torch dialect program to the
+            tcp dialect program (*_tcp.mlir) using `-torch-backend-to-tcp-backend-pipeline`.
+        gen_foo_mlir_llvm:
+            genrule that invokes `tcp-opt` to convert the tcp dialect program to the
+            llvm dialect program (*_llvm.mlir) using `-tcp-to-llvm-pipeline`.
+        gen_foo_llvm_ir:
+            genrule that invokes `mlir-translate` to convert the llvm dialect program to
+            the llvm assembly (*.ll) using `-mlir-to-llvmir`.
+        gen_foo_host_asm:
+            genrule that invokes `llc` on the llvm assembly to generate assembly source
+            (*.S) for the host architecture (CPU).
+        gen_foo_reference_tensors:
+            genrule that invokes `foo_torch_exporter` and saves the reference tensors
+            to a numpy archive (*.npz).
+        gen_foo_execute_test:
+            genrule that invokes `foo_execute_test_generator` to generate a materialized
+            execute_test.cpp for foo.
+
+    The set of auto-generated targets can be obtained by running the following query:
+        bazel query 'attr(name, "foo", //test/AotCompile/...)'
+
     """
+    if not tcp_source and not torch_loader_lib:
+        fail("aot_compile macro requires either `tcp_source` or `torch_loader_lib` " +
+             "to be specified.")
+    if tcp_source and torch_loader_lib:
+        fail("aot_compile macro cannot accept both `tcp_source` and `torch_loader_lib`. " +
+             "Please specify either one.")
+    if torch_loader_lib != None and torch_loader_path == "":
+        fail("aot_compile macro requires `torch_loader_path` to be specified along with " +
+             "`torch_loader_lib`.")
+    if tcp_source and torch_loader_path != "":
+        fail("aot_compile macro cannot accept `torch_loader_path` when `tcp_source` " +
+             "is specified.")
+
+    _name = "_internal_" + name
```

Comment:
https://github.com/cruise-automation/mlir-tcp/commit/ada575353cd2381504447c93bc204773134a9fc2#r139404152

![image](https://github.com/cruise-automation/mlir-tcp/assets/19234106/0ea41584-099b-471e-bd84-e68ec76adc4e)

Reply:
Not worried about conflicts (bazel ensures this at build time) and no strong preference - just carried forward from the previous implementation. We use internal name for the intermediate assets to separate it from the final `.so`.